src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs

```diff
@@ -224,11 +208,17 @@ internal Location CreateLocation(LocationVersionOne v1Location)
                 Message = CreateMessage(message)
             };
 
-            if (string.IsNullOrWhiteSpace(logicalLocationKey) &&
-                !string.IsNullOrWhiteSpace(fullyQualifiedLogicalName))
+            if (!string.IsNullOrWhiteSpace(fullyQualifiedLogicalName))
             {
-                LogicalLocation logicalLocation = CreateLogicalLocation(fullyQualifiedLogicalName);
-                location.FullyQualifiedLogicalName = AddLogicalLocation(logicalLocation);
+                if (string.IsNullOrWhiteSpace(logicalLocationKey))
+                {
+                    LogicalLocation logicalLocation = CreateLogicalLocation(fullyQualifiedLogicalName);
+                    location.FullyQualifiedLogicalName = AddLogicalLocation(logicalLocation);
+                }
+                else
+                {
+                    location.FullyQualifiedLogicalName = logicalLocationKey;
```

Comment:
Bug: In this branch, you don't add the `logicalLocation` object to the dictionary. You might need to refactor `AddLogicalLocation` to separate out the disambiguating logic from the add-to-dictionary logic,

Reply:
I am out of gas, need a glass of wine. I'll finish this later, but don't wait up.

Reply:
No problem. Great progress, almost there.

Reply:
Ah, you are right: Since this overload of `CreateLocation` was called with properties extracted directly from a v1 stack frame, there is already a dictionary entry with the disambiguated key.

That is _so_ worth a comment!

Yes, in this case, we should update the existing v2 dictionary entry with the additional properties (`name`, `fullyQualifiedLogicalName`) -- because in this case, they must both be present, because neither one can possibly match the key. So it's still a bug.

Let me make sure I understand what you mean by this:
> And since we don't get kind nor decoratedName properties from the stackFrame, we will disambiguate if either has a value.

I think you mean that a v1 `logicalLocation` will never equal a v2 `logicalLocation` that has either a `kind` or a `decoratedName`, because a v1 `logicalLocation` doesn't have those properties.

That's true, but is that the right behavior? Suppose you're about to add a `logicalLocation` that matches an existing v2 `logicalLocation` _except_ for `kind` and/or `decoratedName`. _Should_ we assume this is a different logical location? I don't think so. I see that you have a test that verifies this exact behavior, and you'll recall I commented on it, expressing this concern.

Reply:
Ok that's what I was thinking. The statement you ask about is just the condition that made me realize there is a need to pull the existing lL and update it.

Reply:
Ok, so there are two necessary changes:
1. Allow a v1 `logicalLocation` to match a v2 `logicalLocation` that contains `kind` or `decoratedName`. That means you can't just transform the v1 `logicalLocation` to v2 and then use the equality comparer. You'll have to compare a few properties "by hand" -- and boy, will that need a comment.
2. If you find a disambiguated key in the dictionary (the `else` clause), make sure to populate `name` and `fullyQualifiedName` in the `logicalLocation` identified by that key.

Reply:
3. Of course that will break a test, so you'll need to fix the test.

Reply:
OK. (v1 logicalLocation does have kind, so decoratedName is the only issue)

Reply:
You are right. I was inaccurate. v1 `logicalLocation`s do have a `kind`. **BUT!** a `logicalLocation` synthesized from a v1 `stackFrame` -- which is the scenario that this method covers -- will never match a v2 `logicalLocation` with a `kind`, because v1 `stackFrame` doesn't have a `kind`.

We need to fix that and allow a `logicalLocation` synthesized from a v1 `stackFrame` to match an existing v2 `logicalLocation` with a `kind` (or a `decoratedName`). That is my bullet point #&#xFEFF;1 above. Then my #&#xFEFF;2 and #&#xFEFF;3 follow immediately.

Reply:
Check my thinking please: if the stackFrame has a logicalLocationKey, I can just update FQN and Name for that existing item, right? Otherwise, if I have a FQLN, do the create & add with the manual property comparison.

Reply:
Correct. And we still need to do my #&#xFEFF1, #&#xFEFF2, and #&#xFEFF3 above.

Reply:
Sorry, your last sentence implies my #&#xFEFF;1, #&#xFEFF;2, and #&#xFEFF;3. We are on the same page.

Reply:
OK great, thanks